Data/Primitive/PrimArray.hs

```diff
@@ -0,0 +1,488 @@
+{-# INLINE foldrPrimArray #-}
+foldrPrimArray :: forall a b. Prim a => (a -> b -> b) -> b -> PrimArray a -> b
+foldrPrimArray f z arr = go 0
+  where
+    !sz = sizeofPrimArray arr
+    go !i
+      | sz > i = f (indexPrimArray arr i) (go (i+1))
+      | otherwise = z
+
+-- | Strict right-associated fold over the elements of a 'PrimArray'.
+{-# INLINE foldrPrimArray' #-}
+foldrPrimArray' :: forall a b. Prim a => (a -> b -> b) -> b -> PrimArray a -> b
+foldrPrimArray' f z0 arr = go (sizeofPrimArray arr - 1) z0
+  where
+    go !i !acc
+      | i < 0 = acc
+      | otherwise = go (i - 1) (f (indexPrimArray arr i) acc)
+
+-- | Lazy left-associated fold over the elements of a 'PrimArray'.
+{-# INLINE foldlPrimArray #-}
+foldlPrimArray :: forall a b. Prim a => (b -> a -> b) -> b -> PrimArray a -> b
+foldlPrimArray f z arr = go (sizeofPrimArray arr - 1)
+  where
+    go !i
+      | i < 0 = z
+      | otherwise = f (go (i - 1)) (indexPrimArray arr i)
+
+-- | Strict left-associated fold over the elements of a 'PrimArray'.
+{-# INLINE foldlPrimArray' #-}
+foldlPrimArray' :: forall a b. Prim a => (b -> a -> b) -> b -> PrimArray a -> b
+foldlPrimArray' f z0 arr = go 0 z0
+  where
+    !sz = sizeofPrimArray arr
+    go !i !acc
+      | i < sz = go (i + 1) (f acc (indexPrimArray arr i))
+      | otherwise = acc
+
+-- | Strict left-associated fold over the elements of a 'PrimArray'.
+{-# INLINE foldlPrimArrayM' #-}
+foldlPrimArrayM' :: (Prim a, Monad m) => (b -> a -> m b) -> b -> PrimArray a -> m b
+foldlPrimArrayM' f z0 arr = go 0 z0
+  where
+    !sz = sizeofPrimArray arr
+    go !i !acc1
+      | i < sz = do
+          acc2 <- f acc1 (indexPrimArray arr i)
+          go (i + 1) acc2
+      | otherwise = return acc1
+
+-- | Monadic map over a 'PrimArray'.
+{-# INLINE mapPrimArrayM #-}
+mapPrimArrayM :: (Monad m, Prim a, Prim b)
+  => (a -> m b)
+  -> PrimArray a
+  -> m (PrimArray b)
+mapPrimArrayM f arr = go 0 []
+  where
+  !sz = sizeofPrimArray arr
+  go !ix !xs
+    | ix < sz = do
+        !x <- f (indexPrimArray arr ix)
+        go (ix + 1) (x : xs)
+    | otherwise = return (primArrayFromBackwardsListN sz xs)
+
+-- | Monadic map over a 'PrimArray' where the monad allows
+--   primitive actions.
+{-# INLINE mapPrimArrayP #-}
+mapPrimArrayP :: (PrimBase m, Prim a, Prim b)
+  => (a -> m b)
+  -> PrimArray a
+  -> m (PrimArray b)
+mapPrimArrayP = unsafeMapPrimArrayP
+
+-- This function is not exported. It only produces correct
+-- results when the monad is sufficiently affine.
+{-# INLINE unsafeMapPrimArrayP #-}
+unsafeMapPrimArrayP :: (PrimMonad m, Prim a, Prim b)
+  => (a -> m b)
+  -> PrimArray a
+  -> m (PrimArray b)
+unsafeMapPrimArrayP f arr = do
+  let !sz = sizeofPrimArray arr
+  marr <- newPrimArray sz
+  let go !ix = if ix < sz
+        then do
+          b <- f (indexPrimArray arr ix)
+          writePrimArray marr ix b
+          go (ix + 1)
+        else return ()
+  go 0
+  unsafeFreezePrimArray marr
```

Comment:
I think `traverse` rather than `map` explains more clearly what this does.

Reply:
I guess `traverse` should really be preferred to `mapM` when naming things. I'll change this.

Reply:
Actually, I'm conflicted about this. We potentially have four variants:

- mapping purely
- mapping under applicative context
- mapping under monadic context
- mapping under primitive monadic context

Names as it is now: `map`, `traverse`, `mapM`, `mapP`. I could change the latter two to `traverseM` and `traverseP`, but `traverseM` has no precedent.

Reply:
I definitely agree with you on `map` and `traverse`. I'm afraid I don't understand the strictness-based distinction that you draw between the other two. For `PrimArray`, all four of these seem equally strict to me. If a single iteration produces a bottom, the whole thing always bottoms out.

Reply:
You missed it because it's subtle.

```haskell
traverse (maybe (pure undefined) print) $ fromList [Just 1, Nothing, Just 2, Just 3]
```

will print 1, 2, and 3 and then return `undefined`. The monadic versions will print 1 and then throw an exception.

Reply:
Ah, thanks. That makes sense. I had completely missed that. Names could be the following: `map`, `traverse`, `traverse'`, `traverseP'`. Or we could just get rid of the monadic one and call the `PrimBase` one `traverse'`. After all, I think we could expect that someone who needs the extra performance is probably willing to go all the way.

Reply:
We're going to want to make the naming consistent with `Array` and `SmallArray`. Those have *two* notions of strict traversal:

1. Make sure all elements are forced when the resulting array is forced (this takes `Applicative`)
2. Force each element as it is produced (this takes `Monad`).

I'm not entirely sure that no one will want (1). In a sense, (2) seems to be redundant, since you can write

```haskell
foo :: (Traversable t, Monad m) => (a -> m b) -> t a -> m (t b)
foo f = traverse (\a -> f a >>= (pure $!))
```

but you may be able to be more efficient directly.

Reply:
Possibly `unsafeTraversePrimArray'`. But my preference would be to just leave it unexported for the moment and then in a later PR, solve the affine prim monad problem with either a type class or a type family and then weaken the constraint on the `PrimBase` one. Then we wouldn't need the unsafe one.

Reply:
We'd still need the unsafe one. There are contexts where it will be perfectly fine. For example, as long as you're treating `Codensity m` sufficiently abstractly (with `liftCodensity` and `lowerCodensity`), it'll be just fine. I experimented with your way of doing the class, and to my surprise I wasn't able to make the constraints blow up. So we can go with that unless and until someone runs into a performance problem.

Reply:
I've not ever actually used `Codensity`, but I'll take your word on that. For now, I'll go with:

- `mapPrimArray`
- `traversePrimArray` (Applicative)
- `traversePrimArray'` (PrimAffineMonad)
- `unsafeTraversePrimArray'` (PrimMonad)

The latter two functions are actually the same thing. I'll go ahead and add the methodless `PrimAffineMonad` class to make this work.

Reply:
I'm not actually sure that `Codensity` is generally *unsafe* for this (it's a tad confusing), but I know a context where it definitely is safe. Anyway, exposing the unsafe version means people won't need orphan `PrimAffineMonad` instances.